### ISSUE TITLE
Study: HPX, libflatarray

### ISSUE THREAD

**Issue by larryxiao** [larryxiao/libflatarray#1]
- [x] Learn HPX system, build, install and run example.
- [x] Learn LibFlatArray, build and run example.
- [x] Read LibFlatArray code.

**Comment by larryxiao** [comment]
- Reference
  - [HPX | The STE||AR Group](http://stellar-group.org/libraries/hpx/)
  - [STEllAR-GROUP/hpx](https://github.com/STEllAR-GROUP/hpx)
  - [LibFlatArray - Highly Efficient Multi-Dimensional Arrays for C++](http://www.libgeodecomp.org/libflatarray.html)
-   HPX
  -   HPX (High Performance ParalleX) is a general purpose C++ runtime system for parallel and distributed applications of any scale. It strives to provide a unified programming model which transparently utilizes the available resources to achieve unprecedented levels of scalability.  This library strictly adheres to the C++11 Standard and leverages the Boost C++ Libraries which makes HPX easy to use, highly optimized, and very portable.  HPX is developed for conventional architectures including Linux-based systems, Windows, Mac, and the BlueGene/Q, as well as accelerators such as the Xeon Phi.
  -   High Performance ParalleX is the first open-source implementation of the ParalleX execution model.   This model focuses on overcoming the four main barriers to scalability:
    -   Starvation- insufficient concurrent work available to maintain high utilization of all resources.
    -   Latencies- the time-distance delay intrinsic to accessing remote resources and services.
    -   Overhead- the work required for the management of parallel actions and resources on the critical execution path which is not necessary in a sequential variant.
    -   Waiting for contention resolution- the delay due to the lack of availability of oversubscribed shared resources.
  -   In order to overcome these challenges, HPX utilizes these governing principals: latency hiding vs latency avoidance, fine-grained parallelism vs heavyweight threads, constraint based synchronization vs global barriers, adaptive locality control vs static data distribution, work following data vs data following work, and message driven vs message passing.  These principals are not in themselves new; however, the novel combination of all of these concepts is what makes HPX revolutionary.
  -   [Docs | The STE||AR Group](http://stellar-group.org/libraries/hpx/docs/)
-   LibFlatArray
  - Who should use it?
    - Basically anyone who wants to store objects or structs in an array with 2 or more dimensions and needs to iterate through them. Applications include computer simulations (e.g. stencil codes), image processing or linear algebra with complex numbers.
  -   Features
    -   object-oriented interface, but SoA memory layout
    -   good for vectorization (SSE, AVX, CUDA)
    -   zero address generation overhead
    -   fast iterators, yield access to neighboring elements
    -   efficient copy in/out
  -   The Problem with Traditional Arrays & How LibFlatArray Works
    -   ![image](https://cloud.githubusercontent.com/assets/3643298/6879271/ac89f56a-d52b-11e4-98d3-2393454010b2.png)
    -   ![image](http://www.libgeodecomp.org/images/struct_of_arrays.png)

**Comment by larryxiao** [comment]
- short_vec is a primitive data structure that aims to utilize the vectorization capabilities of different architectures
  - it's implementation involves doing contructor, basic arithmetic operations (+-*/), and sqrt operation etc. in vectorized manner
